com.unity.shadergraph/Editor/Generation/Targets/BuiltIn/Editor/AssetPostProcessors/MaterialPostprocessor.cs

```diff
@@ -0,0 +1,138 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using UnityEditor.Rendering.BuiltIn;
+using UnityEngine;
+using static UnityEditor.Rendering.BuiltIn.ShaderUtils;
+
+namespace UnityEditor.Rendering.BuiltIn
+{
+    class MaterialModificationProcessor : AssetModificationProcessor
+    {
+        static void OnWillCreateAsset(string asset)
+        {
+            if (!asset.ToLowerInvariant().EndsWith(".mat"))
+            {
+                return;
+            }
+            MaterialPostprocessor.s_CreatedAssets.Add(asset);
+        }
+    }
+
+    class MaterialReimporter : Editor
+    {
+        // Currently this is never called because there's no way for built-in target shader graph
+        // materials to track when they need to be upgraded at a global level. To do this currently
+        // we'd have to iterate over all materials to see if they need upgrading. We may want to add a
+        // global settings object like UniversalProjectSettings but for built-in to track this.
+        static void ReimportAllMaterials()
+        {
+            string[] guids = AssetDatabase.FindAssets("t:material", null);
+            // There can be several materials subAssets per guid ( ie : FBX files ), remove duplicate guids.
+            var distinctGuids = guids.Distinct();
+
+            int materialIdx = 0;
+            int totalMaterials = distinctGuids.Count();
+            foreach (var asset in distinctGuids)
+            {
+                materialIdx++;
+                var path = AssetDatabase.GUIDToAssetPath(asset);
+                EditorUtility.DisplayProgressBar("Material Upgrader re-import", string.Format("({0} of {1}) {2}", materialIdx, totalMaterials, path), (float)materialIdx / (float)totalMaterials);
+                AssetDatabase.ImportAsset(path);
+            }
+            EditorUtility.ClearProgressBar();
+
+            MaterialPostprocessor.s_NeedsSavingAssets = true;
+        }
+
+        [InitializeOnLoadMethod]
+        static void RegisterUpgraderReimport()
+        {
+        }
+    }
+
+    class MaterialPostprocessor : AssetPostprocessor
+    {
+        public static List<string> s_CreatedAssets = new List<string>();
+        internal static List<string> s_ImportedAssetThatNeedSaving = new List<string>();
+        internal static bool s_NeedsSavingAssets = false;
+
+        internal static readonly Action<Material, ShaderID>[] k_Upgraders = {};
+
+        static void OnPostprocessAllAssets(string[] importedAssets, string[] deletedAssets, string[] movedAssets, string[] movedFromAssetPaths)
```

Comment:
This is here to handle the fixup work needed to make material property overrides work. The material inspector displays some properties like "BlendMode" as an enum that it has to then convert to the actual blend states on the shader. If a new material is added it won't have the correct values configured until it's inspected the first time without this. This is the same as what URP currently does. @cdxntchou can correct me if I'm wrong, but I believe this is also intended to be temporary while we wait for proper asset dependency tracking to be completed.

Reply:
Well, any asset postprocessor has a:

1) Really huge one-time cost. Just because you added a new postprocessor, this will reimport _whole project_ as soon as you upgrade to SG package with this code -- this can easily mean dozens or hundreds of hours in a semi-large production :/

2) Yet another callback that is called after each and every asset import happens, when what it's actually interested in is like 0.0001% of assets, statistically speaking. Is there really no way around the issue without a postprocessor?

like, this bit `If a new material is added it won't have the correct values configured` -- doesn't a new material get default values as they are specified in a shader? For example if I have some sort of `[Enum] _Foobar ("Foobar", Int) = 7` then default value will be 7 for a newly created material. Or am I misunderstanding the problem?

Reply:
Ah right, I did not think about keywords. Oh well, sad that we don't have a better way of handling this yet. Looking forward to when a postprocessor will no longer be needed :)